libs/@blockprotocol/graph/src/stdlib/interval.ts

```diff
@@ -0,0 +1,333 @@
+   Interval B |  [-----]    |    [-----]  |         [---) | (-----]         |   [-----]
+   -----------|-------------|-------------|---------------|-----------------|-------------
+   Merge      |  [-------]  |  [-------]  | [-----------) | (-------------] | [---------]
+   */
+  return {
+    start:
+      compareBounds(lhs.start, rhs.start, "start", "start") <= 0
+        ? lhs.start
+        : rhs.start,
+    end: compareBounds(lhs.end, rhs.end, "end", "end") >= 0 ? lhs.end : rhs.end,
+  } as MergeReturn<LhsInterval, RhsInterval>;
+};
+
+type UnionReturn<
+  LhsInterval extends NonNullTimeInterval,
+  RhsInterval extends NonNullTimeInterval,
+> =
+  | [MergeReturn<LhsInterval, RhsInterval>]
+  | [LhsInterval, RhsInterval]
+  | [RhsInterval, LhsInterval];
+
+/**
+ * Given two {@link NonNullTimeInterval}s, this returns a list of non-adjacent, non-overlapping
+ * {@link NonNullTimeInterval}s which span the space spanned by the input intervals.
+ *
+ * In other words, if the intervals _are_ adjacent, or overlap, then this returns the result of calling
+ * {@link intervalMergeWithInterval} on the intervals, otherwise it returns the two intervals back.
+ *
+ * @param {NonNullTimeInterval}  lhs
+ * @param {NonNullTimeInterval}  rhs
+ */
+export const intervalUnionWithInterval = <
+  LhsInterval extends NonNullTimeInterval = NonNullTimeInterval,
+  RhsInterval extends NonNullTimeInterval = NonNullTimeInterval,
+>(
+  lhs: LhsInterval,
+  rhs: RhsInterval,
+): UnionReturn<LhsInterval, RhsInterval> => {
+  /*
+   Examples   |      1      |      2      |       3       |        4        |      5
+   ===========|=============|=============|===============|=================|=============
+   Interval A |    [-----]  |  [-----]    | [---]         |         [-----] | [---------]
+   Interval B |  [-----]    |    [-----]  |         [---) | (-----]         |   [-----]
+   -----------|-------------|-------------|---------------|-----------------|-------------
+   Union      |  [-------]  |  [-------]  | [---]   [---) | (-----] [-----] | [---------]
+   */
+  if (
+    intervalOverlapsInterval(lhs, rhs) ||
+    intervalIsAdjacentToInterval(lhs, rhs)
+  ) {
+    return [intervalMergeWithInterval(lhs, rhs)];
+  } else if (compareBounds(lhs.start, rhs.start, "start", "start") < 0) {
+    return [lhs, rhs];
+  } else {
+    return [rhs, lhs];
+  }
+};
+
+/**
+ * Given a collection of {@link NonNullTimeInterval}s, this returns a list of non-adjacent, non-overlapping
+ * {@link NonNullTimeInterval}'s which span the space spanned by the input intervals.
+ *
+ * Conceptually this recursively calls {@link intervalUnionWithInterval} pairwise until all intervals have been unioned
+ * with one another. The space spanned by the result will not necessarily be contiguous (may contain gaps).
+ *
+ * @param {NonNullTimeInterval[]} intervals
+ */
+export const unionOfIntervals = <IntervalsType extends NonNullTimeInterval>(
+  ...intervals: IntervalsType[]
```

Comment:
Instead of using an array, does TS has something like a BTreeSet we can utilize here instead?

Reply:
To my knowledge, nothing like you're hoping for with a Rust parallel where you can define custom hash or comparison operators. I also don't think it'd necessarily be faster, arrays are fairly fast and we're only sorting it once.